### DISCUSSION TITLE
Is it possible to fire up VisiData within Python?

### DISCUSSION THREAD

**Discussion by Shahin-rmz** [saulpw/visidata#2026] [Q&A] [answered]
Hi, maybe you know me, because I am the old fan of Visidata from almost day zero.
The question is self explanatory.
I am getting sick of other types of showing dataframes and tabular data in Terminal with python.
Is it possible to open Visidata as python function and show the dataframe as visidata?

## How?
For example Within jupyter cell, having a function, that when we call it, interactive or passive visidata opens up!

**Comment by ajkerrigan** [comment]
I don't know of a smooth way to do this from a jupyter cell :thinking: . Taking Jupyter out of the equation, you can use a flow like this from Python:

```python
import pandas as pd
from visidata import vd

df = ...
vd.view_pandas(df)
```

...but the few times I've tried it from jupyter it hasn't worked well because visidata wants raw terminal access and jupyter is doing its own thing.

One workaround I've only used for example purposes is [passing data into a running vd session](https://asciinema.org/a/590234). I'm assuming that's _probably_ too janky to be a true solution, but maybe still useful in case it triggers any more productive impulses from folks like you who know your way around notebooks.

**Reply by Shahin-rmz** [reply]
Are you sure that the code actually works?
I get this error
```
, in fail(vd, *args)
     69 'Abort with ExpectedException, and display *args* on status as a warning.'
     70 vd.status(*args, priority=2)
---> 71 raise ExpectedException(args[0] if args else '')

ExpectedException: raw() returned ERR
```
I am trying to open it with a loader.

**Reply by ajkerrigan** [reply]
> Are you sure that the code actually works?

Nope. It only earned the "works on my machine" badge.

> I get this error
> ```
> , in fail(vd, *args)
>      69 'Abort with ExpectedException, and display *args* on status as a warning.'
>      70 vd.status(*args, priority=2)
> ---> 71 raise ExpectedException(args[0] if args else '')
> 
> ExpectedException: raw() returned ERR
> ```
> I am trying to open it with a loader.

Right. That's what I meant here:

> ...but the few times I've tried it from jupyter it hasn't worked well because visidata wants raw terminal access and jupyter is doing its own thing.

I don't use jupyter often so didn't have the traceback handy. But it looks like you're hitting the same trouble. Makes sense.

**Comment by Shahin-rmz** [comment]
one year later and I still could not open visidata within my python apps.
is it possible?
@saulpw 
I use vim and  connect it to jupyter console.
As file manager I use vifm and it has a [Plugin](https://github.com/vifm/vifm.vim) that we can fire up vifm inside vim. wanted to know if we can do the same to interact with our dataframes with Visidata inside python or vim.
Thanks

**Reply by saulpw** [reply]
It works fine from the Python console and ipython:

```
from visidata import vd
vd.view('hello')
```

**Reply by Shahin-rmz** [reply]
You are right,
If I run my app with `python3 main.py` it works fine. 
([I use jupyter vim integration](https://github.com/jupyter-vim/jupyter-vim)) it sends codes inside a cell to a ipython instance run it.
in this case I get this error:
![image](https://github.com/user-attachments/assets/a7d36a4e-43eb-4a60-8ccd-35663fa5543d)

**Reply by saulpw** [reply, marked as answer]
Yes, this is the same error @ajkerrigan showed above.  Looks like Jupyter doesn't support curses raw() mode (used for TUIs).  So it's not possible within Jupyter.  Sorry.
